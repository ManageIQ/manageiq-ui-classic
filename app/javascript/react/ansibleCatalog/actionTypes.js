const prefix = '@@manageiq/ansibleCatalogItem/';
export const LOAD_CATALOGS = `${prefix} Load catalogs`;
export const LOAD_DIALOGS = `${prefix} Load dialogs`;
export const LOAD_REPOS = `${prefix} LOAD_REPOS`;
export const LOAD_PLAYBOOKS = `${prefix} LOAD_PLAYBOOKS`;
export const LOAD_MACHINE_CREDENTIALS = `${prefix} LOAD_MACHINE_CREDENTIALS`;
export const LOAD_CLOUD_TYPES = `${prefix} LOAD_CLOUD_TYPES`;
export const LOAD_VAULT_CREDENTIALS = `${prefix} LOAD_VAULT_CREDENTIALS`;
export const LOAD_CLOUD_CREDENTIALS = `${prefix} LOAD_CLOUD_CREDENTIALS`;
export const LOAD_CATALOG_ITEM = `${prefix} LOAD_CATALOG_ITEM`;
export const LOAD_CLOUD_CREDENTIAL = `${prefix} LOAD_CLOUD_CREDENTIAL`;
export const CREATE_CATALOG_ITEM = `${prefix} Create catalog item`;
export const DUPLICATE_DROPDOWNS = `${prefix} DUPLICATE_DROPDOWNS`;
