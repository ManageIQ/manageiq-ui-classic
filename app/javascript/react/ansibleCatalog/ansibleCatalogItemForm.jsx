import React from 'react';
import { Form, Field, FormSpy} from 'react-final-form';
import { Form as PfForm, Button, TabContainer, TabContent, Spinner, TabPane, Nav, NavItem, MessageDialog, Icon } from 'patternfly-react';
import AnsiblePlaybookFields from './ansiblePlaybookFields';
import arrayMutators from 'final-form-arrays';
import FormButtons from '../../forms/form-buttons';
import * as helpers from './helpers';
import {ANSIBLE_FIELDS, DEFAULT_PLACEHOLDER, CATALOG_ITEM_PROPS, DROPDOWN_OPTIONS} from './constants';
import { FinalFormSelect } from '@manageiq/react-ui-components/dist/forms';
import { DialogFields } from './dialogFields';
import {cloneDeep} from 'lodash';

export class AnsibleCatalogItemForm extends React.Component {
constructor(props) {
    super(props);
    const initialValues = helpers.setFormDefaultValues()
    this.state = {
      initialValues,
      formValues: cloneDeep(initialValues),
      isLoading: true,
      isNew: true,
      showCopyProvisionDialog: false,
    }
  }
  componentDidMount(){
    const { loadCatalogs, loadDialogs, loadRepos, loadCredentials, loadCloudTypes,
      loadCatalogItem, loadCloudCredential, region, catalogItemFormId, availableCatalogs } = this.props;
    loadCatalogs(availableCatalogs);
    loadDialogs();
    loadRepos(region);
    loadCredentials('Machine');
    loadCredentials('Vault');
    loadCloudTypes();
    if (catalogItemFormId){
      this.setState({isNew: false});
      loadCatalogItem(catalogItemFormId).then(() => {
        const {ansibleCatalog} = this.props;
        const {initialValues} = this.state;
        const catalogItem = ansibleCatalog.catalogItem;
        loadCloudCredential(catalogItem.config_info.provision.cloud_credential_id).then(() => {
                    ['name','description','long_description','service_template_catalog_id'].forEach((field) => {
                      if (catalogItem[field]){
                        initialValues[field] = catalogItem[field];
                      }
                    });
                    ANSIBLE_FIELDS.forEach((field) => {
                      ['provision', 'retirement'].forEach(prefix => {
                        if (catalogItem.config_info[prefix][field]){
                          initialValues[`${prefix}_${field}`] = catalogItem.config_info[prefix][field];
                        }
                      });
                   });
                   initialValues['provision_extra_vars'] = helpers.formatExtraVars(initialValues['provision_extra_vars']);
                   initialValues['retirement_extra_vars'] = helpers.formatExtraVars(initialValues['retirement_extra_vars']);
                   initialValues.dialog_id = (catalogItem.config_info.provision.dialog_id? catalogItem.config_info.provision.dialog_id: '');
                   initialValues.remove_resources = (catalogItem.config_info.retirement.remove_resources ?
                     catalogItem.config_info.retirement.remove_resources : initialValues.remove_resources);
                   this.setState({initialValues,formValues:cloneDeep(initialValues), isLoading: false});
        });
      });
    } else {
      this.setState({isLoading: false});
    }
  }

  submitForm = (values) => {
    const {isNew} = this.state;
    const {catalogItemFormId} = this.props;
    const { name, description, display, long_description, service_template_catalog_id } = values;
    const provisionFields = {}, retirementFields = {};
    ANSIBLE_FIELDS.forEach((field) => {
      const isIdField = field.includes('_id');
      if (isIdField) { //blank ids passed to api cause failures
        ['provision', 'retirement'].forEach(prefix => {
          const value = values[`${prefix}_${field}`];
          if (value) {
            (prefix === 'provision' ? provisionFields[field] = value: retirementFields[field] = value);
          }
        });
      } else {
        provisionFields[field] = values[`provision_${field}`];
        retirementFields[field] = values[`retirement_${field}`];
      }
    });

    const provision = {
      dialog_id: values['dialog_id'],
      ...provisionFields,
      extra_vars: helpers.buildExtraVarsObj(provisionFields.extra_vars),
    };

    if (values.dialogOption === 'new') {
      provision.new_dialog_name = values.dialog_name;
      delete provision.dialog_id;
    }

    const retirement = {
      ...retirementFields,
      extra_vars: helpers.buildExtraVarsObj(retirementFields.extra_vars),
      remove_resources: values.remove_resources,
    };

    const catalogItem = {
      action: (isNew ? 'create': 'edit'),
      resource: {
        name,
        description,
        long_description,
        display,
        service_template_catalog_id,
        prov_type: 'generic_ansible_playbook',
        type: 'ServiceTemplateAnsiblePlaybook',
        config_info: {
            provision,
            retirement
        }
      }
    };
    helpers.submitCatalogForm(`/api/service_templates/${!isNew? catalogItemFormId:''}`, isNew, catalogItem, catalogItemFormId);
  }

  closeCopyProvisionDialog = () => {
    this.setState({showCopyProvisionDialog: false});
  }
  copyFormValues = (args, state, utils) => {
    ANSIBLE_FIELDS.forEach((field) => {
      utils.changeValue(state, `retirement_${field}`, () => state.formState.values[`provision_${field}`])
    });
    this.props.duplicateDropdowns(['playbooks', 'cloudCredentials']);
    this.closeCopyProvisionDialog();
  }
  handleCancel = () => {
    const message = (this.state.isNew ? __('Add of Catalog Item was cancelled by the user') :
      __('Edit of Catalog Item was cancelled by the user'));
    miqFlashLater({ message });
    window.location.href = '/catalog/explorer/';
  };
  handleResetClick = (reset) => {
    const {initialValues} = this.state;
    window.add_flash( __('All changes have been reset'),'warn');
    reset(initialValues);
    this.setState({extraVarsChanged: false,isLoading:true, formValues: cloneDeep(initialValues)}, this.reloadForm);
  }
  reloadForm=() =>{
    this.setState({isLoading: false});
  }

  render() {
    const { region, ansibleCatalog } = this.props;
    const { isLoading, isNew } = this.state;
    const dropdowns = ansibleCatalog.dropdowns || {};
    const formFields = helpers.getAnsibleCatalogItemFields(dropdowns);
    const formMutators = { ...arrayMutators, copyValues: this.copyFormValues };
    return (
      <div>
        <div id="flash_msg_div" />
        {(isLoading ? (<Spinner loading size="lg" />) :
          (<Form
            initialValues={this.state.formValues}
            onSubmit={this.submitForm}
            mutators={{
              ...formMutators
            }}
            render={({ invalid,
              values,
              submitting,
              pristine,
              dirty,
              handleSubmit,
              form: { change, reset, mutators: { push, copyValues } } }) => (
                <PfForm horizontal onSubmit={handleSubmit}>
                  <MessageDialog
                    onHide={this.closeCopyProvisionDialog}
                    show={this.state.showCopyProvisionDialog}
                    primaryAction={copyValues}
                    primaryActionButtonContent={__('Copy')}
                    secondaryAction={this.closeCopyProvisionDialog}
                    secondaryActionButtonContent={__('Cancel')}
                    title={__('Copy from Provisioning')}
                    icon={(<Icon type="pf" name="warning-triangle-o" />)}
                    primaryContent={helpers.provisionDialogMessage()}
                  />
                  {
                    formFields.map(fields => (helpers.renderFormField(fields, values)))
                  }
                  <TabContainer defaultActiveKey={1} id="ansibleCatalogForm">
                    <div>
                      <Nav bsClass={'nav nav-tabs nav-tabs-pf'}>
                        <NavItem eventKey={1}>{__('Provisioning')}</NavItem>
                        <NavItem eventKey={2}>{__('Retirement')}</NavItem>
                      </Nav>
                      <TabContent animation>
                        <TabPane eventKey={1} mountOnEnter={true} unmountOnExit={true}>
                          <AnsiblePlaybookFields
                            region={region}
                            prefix='provision'
                            addExtraVars={push}
                            initialValues={this.state.initialValues}
                            formValues={values}
                            changeField={change}
                          >
                            <DialogFields dropdownOptions={dropdowns.dialogs} />
                          </AnsiblePlaybookFields>
                        </TabPane>
                        {/* Retirement tab */}
                        <TabPane eventKey={2} mountOnEnter={true}>
                          <div>
                            <div className="col-xs-1" />
                            <div className="col-xs-11 provisioning-copy-button">
                              <Button
                                type="button"
                                bsStyle="link"
                                onClick={() => this.setState({ showCopyProvisionDialog: true })}>
                                <Icon type="fa" name="paste" />
                                &nbsp;
                          {__('Copy from Provisioning')}
                              </Button>
                            </div>
                          </div>
                          <AnsiblePlaybookFields
                            region={region}
                            prefix='retirement'
                            addExtraVars={push}
                            initialValues={this.state.initialValues}
                            formValues={values}
                            changeField={change}
                          >
                            <Field
                              name='remove_resources'
                              component={FinalFormSelect}
                              label={__('Remove Resources?')}
                              options={helpers.getResourceOptions(values)}
                              placeholder={DEFAULT_PLACEHOLDER}
                            />
                          </AnsiblePlaybookFields>
                        </TabPane>
                      </TabContent>
                    </div>
                  </TabContainer>
                  <FormButtons
                    newRecord={isNew}
                    saveable={(!invalid && !submitting && dirty) || this.state.extraVarsChanged}
                    saveClicked={handleSubmit}
                    addClicked={handleSubmit}
                    cancelClicked={this.handleCancel}
                    resetClicked={() => { this.handleResetClick(reset); }}
                    pristine={pristine}
                  />
                </PfForm>
              )}
          />
          )
        )}
      </div>
    );
  }
}

AnsibleCatalogItemForm.propTypes = CATALOG_ITEM_PROPS;
AnsibleCatalogItemForm.defaultProps = {
  catalogItemFormId: null
};
export default AnsibleCatalogItemForm;
