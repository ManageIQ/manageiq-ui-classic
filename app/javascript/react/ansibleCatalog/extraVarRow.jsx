import React from 'react';
import PropTypes from 'prop-types';
import { ExtraVarField } from './extraVarField';
import { Button } from 'patternfly-react';

export class ExtraVarRow extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      isEditing: false,
      row: {key: '', default: '',},
    };
  }
  componentDidMount() {
    const {values, fieldIndex} = this.props;
    if (!values[fieldIndex]) {
      this.setState({isEditing: true});
    } else {
      this.setState({row: values[fieldIndex]});
    }
  }
  componentDidUpdate(prevState) {
    const {fieldIndex, values} = this.props;
    const {isEditing} = this.state;
    if (prevState.values[fieldIndex] !== values[fieldIndex] && !isEditing) {
      this.setState({row: values[fieldIndex]});
    }
  }
  updateField = () => {
    let {isEditing, row} = this.state;
    const {changeField, values, fieldIndex, extraVarsField} = this.props;
    if (isEditing && (!row.key || !row.default)) {
      return;
    }
    isEditing = !isEditing;
    values[fieldIndex] = {...row}
    changeField(extraVarsField, values);
    this.setState({isEditing});
  }
  revertField() {
    const {row} = this.state;
    const  {values, fieldIndex} = this.props;
    if(!row.key) {
     return this.removeField();
    }
    else {
      this.setState({ isEditing: false, row: values[fieldIndex] });
    }
  }
  removeField() {
    const {fieldIndex, removeField} = this.props;
    removeField(fieldIndex);
  }
  renderButton(defaultIcon, editIcon, onclick) {
    const { isEditing } = this.state;
    return (<Button
      type="button"
      bsStyle="link"
      onClick={onclick}
    >
      <i className={`pficon ${(isEditing ? editIcon : defaultIcon)}`} />
    </Button>);
  }
  handleRemoveClick = () => {
    const { isEditing } = this.state;
    if (isEditing) {
      this.revertField();
    } else {
      this.removeField();
    }
  }
  handleFieldUpdate = event => {
    const {row} = this.state;
    row[event.target.name] = event.target.value;
    this.setState({row});
  }
  render() {
    const {isEditing, row} = this.state;
    const {fieldName, values, fieldIndex} = this.props;
    const currentExtraVar = values[fieldIndex] || [];
    const textFields = [
      { fieldName: 'key', label: __('Variable')},
      { fieldName: 'default', label: __('Default')},
    ];
    if (!row && !isEditing) { //this happens because of these fields being part of an array
      return null;
    }
    return (
      <tr>
      { textFields.map((field,index) => {
        const currentValue = (currentExtraVar ? currentExtraVar[field.fieldName] : null);
        const showError = !row[field.fieldName]? 'has-error': '';
        return (
          <td key={index} className={showError}>
          {isEditing &&
          <input type="text" className="form-control" name={field.fieldName} defaultValue={currentValue} onChange={this.handleFieldUpdate}/>
          }
          {isEditing && !row[field.fieldName] &&
                <div className="help-block">
                  {__('Required')}
                </div>
          }
            <ExtraVarField field={{name:fieldName,
            label: field.label,
            isEditing, fieldName: field.fieldName, value: currentValue, }} />
          </td>);
      })
      }
      <td>
        { this.renderButton('pficon-edit','pficon-save',this.updateField) }
        { this.renderButton('pficon-delete','pficon-close',this.handleRemoveClick) }
      </td>
    </tr>);
  }
}
ExtraVarRow.propTypes = {
  changeField: PropTypes.func,
  removeField: PropTypes.func,
  fieldIndex: PropTypes.number,
  values: PropTypes.array,
  fieldName: PropTypes.string,
  extraVarsField: PropTypes.string,
};

export default ExtraVarRow;
