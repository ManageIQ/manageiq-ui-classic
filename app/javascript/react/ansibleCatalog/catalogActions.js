import * as actionTypes from './actionTypes';
import { API } from '../../http_api';

export const commonParams = '&expand=resources&sort_by=name&sort_order=ascending';
export const credentialsRequest = type => `/api/authentications?collection_class=ManageIQ::Providers::EmbeddedAnsible::AutomationManager::${type}&attributes=id,name${commonParams}`;
export const catchError = (error) => {
  if (!error.data) {
    error.data = { // eslint-disable-line no-param-reassign
      error: { message: `Request failed, ${error.message}` },
    };
  }
  window.add_flash(error.data.error.message, 'error');
};
export const loadCatalogs = catalogs => ({
  type: actionTypes.LOAD_CATALOGS,
  payload: catalogs,
});
export const loadDialogs = () => dispatch => API.get('/api/service_dialogs/?expand=resources&attributes=id,label&sort_by=label&sort_order=ascending').then((data) => {
  dispatch({
    type: actionTypes.LOAD_DIALOGS,
    payload: data,
  });
}).catch((error) => {
  catchError(error);
});
export const loadRepos = region => dispatch => API.get(`/api/configuration_script_sources?collection_class=ManageIQ::Providers::EmbeddedAnsible::AutomationManager::ConfigurationScriptSource&attributes=id,name&filter[]=region_number=${region}${commonParams}`).then((data) => {
  dispatch({
    type: actionTypes.LOAD_REPOS,
    payload: data,
  });
}).catch((error) => {
  catchError(error);
});
export const loadPlaybooks = (region, repoId, playbookType) => dispatch => API.get(`/api/configuration_script_sources/${repoId}/configuration_script_payloads?filter[]=region_number=${region}${commonParams}`).then((data) => {
  dispatch({
    type: actionTypes.LOAD_PLAYBOOKS,
    payload: { ...data, playbookType },
  });
}).catch((error) => {
  catchError(error);
});
export const loadCredentials = type => (dispatch) => {
  const actionType = (type === 'Machine' ? actionTypes.LOAD_MACHINE_CREDENTIALS : actionTypes.LOAD_VAULT_CREDENTIALS);

  return API.get(credentialsRequest(`${type}Credential`)).then((data) => {
    dispatch({
      type: actionType,
      payload: data,
    });
  }).catch((error) => {
    catchError(error);
  });
};
export const loadCloudTypes = () => dispatch => API.options('/api/authentications').then((data) => {
  dispatch({
    type: actionTypes.LOAD_CLOUD_TYPES,
    payload: data.data,
  });
}).catch((error) => {
  catchError(error);
});
export const loadCloudCredentials = (fieldType, cloudType) => dispatch => API.get(`/api/authentications?collection_class=${cloudType}&attributes=id,name${commonParams}`).then((data) => {
  dispatch({
    type: actionTypes.LOAD_CLOUD_CREDENTIALS,
    payload: { ...data, fieldType },
  });
}).catch((error) => {
  catchError(error);
});
export const loadCatalogItem = catalogItemId => dispatch => API.get(`/api/service_templates/${catalogItemId}`).then((data) => {
  dispatch({
    type: actionTypes.LOAD_CATALOG_ITEM,
    payload: data,
  });
}).catch((error) => {
  catchError(error);
});
export const loadCloudCredential = credentialId => dispatch => API.get(`/api/authentications/${credentialId}`).then((data) => {
  dispatch({
    type: actionTypes.LOAD_CLOUD_CREDENTIAL,
    payload: data,
  });
}).catch((error) => {
  catchError(error);
});
export const duplicateDropdowns = fields => ({
  type: actionTypes.DUPLICATE_DROPDOWNS,
  payload: fields,
});
