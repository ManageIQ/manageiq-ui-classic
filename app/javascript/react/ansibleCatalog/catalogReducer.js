import sortBy from 'lodash/sortBy';
import * as actionTypes from './actionTypes';
import { buildDropDown } from './helpers';

const mapCloudType = (cloudTypes) => {
  const credentialTypes = cloudTypes.embedded_ansible_credential_types;
  const cloudTypeList = [];
  Object.keys(credentialTypes).forEach((key) => {
    if (credentialTypes[key].type === 'cloud') {
      cloudTypeList.push({
        label: credentialTypes[key].label,
        value: key,
      });
    }
  });
  return sortBy(cloudTypeList, 'label');
};

const initialState = {
  loading: false,
  dropdowns: {},
};

export default (state = initialState, action) => {
  const dropdowns = { ...state.dropdowns };
  const updatedState = { ...state };

  switch (action.type) {
    case actionTypes.LOAD_CATALOGS:
      dropdowns.catalogs = buildDropDown(action.payload, 0, 1);
      break;
    case actionTypes.LOAD_DIALOGS:
      dropdowns.dialogs = buildDropDown(action.payload.resources, 'label', 'id');
      break;
    case actionTypes.LOAD_REPOS:
      dropdowns.repos = buildDropDown(action.payload.resources, 'name', 'id');
      break;
    case actionTypes.LOAD_PLAYBOOKS: {
      const playbookDropDown = `${action.payload.playbookType}_playbooks`;
      dropdowns[playbookDropDown] = buildDropDown(action.payload.resources, 'name', 'id');
      break;
    }
    case actionTypes.LOAD_MACHINE_CREDENTIALS:
      dropdowns.machineCredentials = buildDropDown(action.payload.resources, 'name', 'id');
      break;
    case actionTypes.LOAD_VAULT_CREDENTIALS:
      dropdowns.vaultCredentials = buildDropDown(action.payload.resources, 'name', 'id');
      break;
    case actionTypes.LOAD_CLOUD_TYPES:
      dropdowns.cloudTypes = mapCloudType(action.payload.credential_types);
      break;
    case actionTypes.LOAD_CLOUD_CREDENTIALS: {
      const { fieldType } = action.payload;
      dropdowns[`${fieldType}_cloudCredentials`] = buildDropDown(action.payload.resources, 'name', 'id');
      break;
    }
    case actionTypes.LOAD_CATALOG_ITEM:
      updatedState.catalogItem = action.payload;
      break;
    case actionTypes.LOAD_CLOUD_CREDENTIAL:
      updatedState.cloudCredential = action.payload;
      break;
    case actionTypes.DUPLICATE_DROPDOWNS:
      action.payload.forEach((field) => {
        dropdowns[`retirement_${field}`] = dropdowns[`provision_${field}`];
      });
      break;
    default:
      return updatedState;
  }
  updatedState.dropdowns = dropdowns;
  return updatedState;
};
