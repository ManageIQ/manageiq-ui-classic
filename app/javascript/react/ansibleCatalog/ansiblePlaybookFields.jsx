/* eslint-disable jsx-a11y/label-has-for */
import React from 'react';
import { Field } from 'react-final-form';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { required } from 'redux-form-validators';
import { FieldArray } from 'react-final-form-arrays';
import { Button } from 'patternfly-react';
import { FinalFormSelect } from '@manageiq/react-ui-components/dist/forms';
import * as catalogActions from './catalogActions';
import { ExtraVarRow } from './extraVarRow';
import { DEFAULT_PLACEHOLDER, ANSIBLE_PLAYBOOK_FIELDS_PROPS } from './constants';
import { renderFormField, getAnsiblePlaybookFields } from './helpers';

export class AnsiblePlaybookFields extends React.Component {
  constructor(props) {
    super(props);

    const { formValues, prefix } = this.props;
    this.state = {
      isProvisioning: prefix === 'provision',
      repoId: formValues[`${prefix}_repository_id`],
    };
  }
  componentDidMount() {
    const { formValues, prefix } = this.props;

    if (formValues[`${prefix}_cloud_type`]) {
      this.updateCloudCredentialsList(formValues[`${prefix}_cloud_type`]);
    }
    if (this.state.repoId) {
      this.updatePlaybookList();
    }
  }

  updatePlaybookList() {
    const { region, loadPlaybooks, prefix } = this.props;
    if (!this.state.repoId) {
      return;
    }
    loadPlaybooks(region, this.state.repoId, prefix);
  }
  updateCloudCredentialsList(cloudType) {
    this.props.loadCloudCredentials(this.props.prefix, cloudType);
  }

  addExtraVarToState() {
    this.props.addExtraVars(`${this.props.prefix}_extra_vars`, undefined);
  }

  render() {
    const {
      prefix, data, changeField, children, formValues,
    } = this.props;
    const { isProvisioning } = this.state;
    const dropdowns = data.dropdowns || {};
    const extraVarsField = `${prefix}_extra_vars`;
    const formFields = getAnsiblePlaybookFields(this.props, this.state.isProvisioning);
    const { initialValues } = this.props;
    return (
      <div className="row">
        <div className="col-md-12 col-lg-6">
          <Field
            label={__('Repository')}
            name={`${prefix}_repository_id`}
            validateOnMount={isProvisioning}
            validate={isProvisioning ? required({ msg: __('Required') }) : undefined}
            render={({ input, meta }) => (
              <FinalFormSelect
                inputColumnSize={8}
                labelColumnSize={3}
                name={`${prefix}_repository_id`}
                searchable
                options={dropdowns.repos || []}
                placeholder={DEFAULT_PLACEHOLDER}
                validateOnMount={isProvisioning}
                validate={isProvisioning ? required({ msg: __('Required') }) : undefined}
                label={__('Repository')}
                input={{
                  ...input,
                  onChange: (value) => {
                    if (value) {
                      this.setState({ repoId: value }, this.updatePlaybookList);
                      changeField(`${prefix}_playbook_id`, '');
                    }
                    input.onChange(value);
                  },
                }}
                meta={meta}
              />)}
          />
          { formFields.map(field => renderFormField(field, formValues)) }
          <Field
            name={`${prefix}_cloud_type`}
            label={__('Cloud Type')}
            render={({ input, meta }) => (
              <FinalFormSelect
                inputColumnSize={8}
                labelColumnSize={3}
                name={`${prefix}_cloud_type`}
                options={dropdowns.cloudTypes || []}
                label={__('Cloud Type')}
                placeholder={DEFAULT_PLACEHOLDER}
                input={{
                  ...input,
                  onChange: (value) => {
                    if (value) {
                      this.updateCloudCredentialsList(value);
                    }
                    input.onChange(value);
                  },
                }}
                meta={meta}
              />)}
          />
          {
            (formValues[`${prefix}_cloud_type`] !== '') &&
            <Field
              inputColumnSize={8}
              labelColumnSize={3}
              name={`${prefix}_cloud_credential_id`}
              label={__('Cloud Credentials')}
              component={FinalFormSelect}
              options={dropdowns[`${prefix}_cloudCredentials`] || []}
              placeholder={DEFAULT_PLACEHOLDER}
            />
          }
        </div>
        <div className="col-md-12 col-lg-6">
          <div className="form-group">
            <div className="col-xs-2">
              <label className="control-label">{__('Variables & Default Values')}</label>
            </div>
            <div className="col-xs-8">
              <div>
                <Button
                  type="button"
                  bsStyle="primary"
                  onClick={() => { this.addExtraVarToState(); }
                  }
                >
                  {__('Add')}
                </Button>
              </div>
              <table className="table table-bordered table-striped">
                <thead>
                  {formValues[extraVarsField].length > 0 &&
                  <tr>
                    <th>{__('Variable')}</th>
                    <th>{__('Default')}</th>
                    <th>{__('Action')}</th>
                  </tr>
                  }
                </thead>
                <tbody>
                  <FieldArray
                    name={extraVarsField}
                    isEqual={(prevValues, newValues) => {
                      const fieldValuesChanged = JSON.stringify(formValues[extraVarsField]) !== JSON.stringify(initialValues[extraVarsField]);
                      const fieldSizeChanged = prevValues.length !== newValues.length;
                      return !fieldValuesChanged && !fieldSizeChanged;
                    }}
                  >
                    {({ fields }) =>
                      fields.map((name, index) => (
                        <ExtraVarRow
                          fieldName={name}
                          key={name}
                          fieldIndex={index}
                          removeField={fields.remove}
                          changeField={changeField}
                          extraVarsField={extraVarsField}
                          values={formValues[extraVarsField]}
                        />
                      ))}
                  </FieldArray>
                </tbody>
              </table>
            </div>
          </div>
          {children}
        </div>
      </div>
    );
  }
}

AnsiblePlaybookFields.propTypes = ANSIBLE_PLAYBOOK_FIELDS_PROPS;
const mapStateToProps = state => ({
  data: state.ansibleCatalog || {},
});
const actions = { ...catalogActions };
const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(AnsiblePlaybookFields);
