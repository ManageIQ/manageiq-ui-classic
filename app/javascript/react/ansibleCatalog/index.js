import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import * as catalogActions from './catalogActions';

import reducer from './catalogReducer';
import { AnsibleCatalogItemForm } from './ansibleCatalogItemForm';

const actions = { ...catalogActions };
const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch);
const mapStateToProps = state => ({
  ansibleCatalog: state.ansibleCatalog || {},
});
ManageIQ.redux.addReducer({ ansibleCatalog: reducer });
export default connect(mapStateToProps, mapDispatchToProps)(AnsibleCatalogItemForm);
