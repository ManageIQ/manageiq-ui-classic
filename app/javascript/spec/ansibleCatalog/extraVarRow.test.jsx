import React from 'react';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { ExtraVarRow } from '../../react/ansibleCatalog/extraVarRow';

describe('ExtraVarRow component', () => {
  const initialProps = {
    changeField: jest.fn(),
    removeField: jest.fn(),
    fieldIndex: 1,
    values: [
      { key: 'field1', default: '' },
      { key: 'field2', default: '' },
    ],
    fieldName: 'provisionExtraVars',
    extraVarsField: 'provisionExtraVars[0]',
  };

  it('should render correctly', () => {
    const wrapper = mount(<ExtraVarRow {...initialProps} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should allow a field to update', () => {
    const wrapper = shallow(<ExtraVarRow {...initialProps} />);
    const instance = wrapper.instance();
    instance.setState({
      isEditing: true,
      row: {
        key: 'test',
        default: 'test',
      },
    });
    instance.updateField();

    expect(instance.state.isEditing).toBeFalsy();
  });

  it('should allow a field to be deleted', () => {
    const removeField = jest.fn();
    const props = { ...initialProps, removeField };
    const wrapper = shallow(<ExtraVarRow {...props} />);
    const instance = wrapper.instance();
    instance.removeField();
    expect(removeField).toHaveBeenCalledWith(1);
  });

  it('should allow a button to be rendered', () => {
    const wrapper = mount(<ExtraVarRow {...initialProps} />);
    const button = wrapper.instance().renderButton('pficon-edit', 'pficon-save', jest.fn());
    const buttonWrapper = mount(button);
    expect(toJson(buttonWrapper)).toMatchSnapshot();
  });

  it('should handle when the remove button is clicked', () => {
    const wrapper = mount(<ExtraVarRow {...initialProps} />);
    const instance = wrapper.instance();
    instance.setState({ isEditing: true });
    const revertMethod = jest.fn();
    instance.revertField = revertMethod;
    wrapper.update();
    instance.handleRemoveClick();
    expect(revertMethod).toHaveBeenCalled();
  });
});
