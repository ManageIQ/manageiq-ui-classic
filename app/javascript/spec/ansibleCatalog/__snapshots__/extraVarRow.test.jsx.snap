// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExtraVarRow component should allow a button to be rendered 1`] = `
<Button
  active={false}
  block={false}
  bsClass="btn"
  bsStyle="link"
  disabled={false}
  onClick={[MockFunction]}
  type="button"
>
  <button
    className="btn btn-link"
    disabled={false}
    onClick={[MockFunction]}
    type="button"
  >
    <i
      className="pficon pficon-edit"
    />
  </button>
</Button>
`;

exports[`ExtraVarRow component should render correctly 1`] = `
<ExtraVarRow
  changeField={[MockFunction]}
  extraVarsField="provisionExtraVars[0]"
  fieldIndex={1}
  fieldName="provisionExtraVars"
  removeField={[MockFunction]}
  values={
    Array [
      Object {
        "default": "",
        "key": "field1",
      },
      Object {
        "default": "",
        "key": "field2",
      },
    ]
  }
>
  <tr>
    <td
      className=""
      key="0"
    >
      <Component
        field={
          Object {
            "fieldName": "key",
            "isEditing": false,
            "label": "Variable",
            "name": "provisionExtraVars",
            "value": "field2",
          }
        }
      >
        <div>
          field2
        </div>
      </Component>
    </td>
    <td
      className="has-error"
      key="1"
    >
      <Component
        field={
          Object {
            "fieldName": "default",
            "isEditing": false,
            "label": "Default",
            "name": "provisionExtraVars",
            "value": "",
          }
        }
      >
        <div />
      </Component>
    </td>
    <td>
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="link"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <button
          className="btn btn-link"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <i
            className="pficon pficon-edit"
          />
        </button>
      </Button>
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="link"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <button
          className="btn btn-link"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <i
            className="pficon pficon-delete"
          />
        </button>
      </Button>
    </td>
  </tr>
</ExtraVarRow>
`;
